### ISSUE TITLE
Ignore synthethic locations in visitor

### ISSUE THREAD

**Issue by JonathanStarup** [flix/flix#9482]
Follow up to #8935 and solves #8935

For resolver generated partial applications to work with hover, we need to 1) recurse into synthetic locations but 2) not consume them. 2) is the missing part

**Comment by magnus-madsen** [comment]
I think it is more complicated than that.

We need an invariant that says when to visit/not to visit a synthetic location and when to recurse. 

Do you have a suggestion?

**Comment by JonathanStarup** [comment]
I think that it sounds like our use or synthetic/real is wrong then. Why should lsp things care about compiler generated stuff?

**Comment by magnus-madsen** [comment]
> I think that it sounds like our use of synthetic/real is wrong then. Why should lsp things care about compiler generated stuff?

The problem is all our desugarings etc.

**Comment by JonathanStarup** [comment]
yeah that makes sense, but  the AST should know whether its synthetic or not, otherwise the visitor has the impossible task of undoing the desugaring that happened

**Comment by magnus-madsen** [comment]
> yeah that makes sense, but the AST should know whether its synthetic or not, otherwise the visitor has the impossible task of undoing the desugaring that happened

Agreed. But I don't know how to solve it. Consider par, forM, etc. What should the AST look like and what information should it carry?

**Comment by JonathanStarup** [comment]
yeah for those its not easy. i guess the best solution in this perspective is to cut down on desugaring. i remember you also saying that a while ago, do you still feel like thats the direction it should go?

**Comment by magnus-madsen** [comment]
I think we need come up up a plan like this:

Write a verifier that enforces:

- Source locations properly nested within each other
- ??? invariant about synthetic locations.

Even if we "drop desugar" there are still many pieces of the AST that are synthetic. Just small things here and there.

**Comment by JonathanStarup** [comment]
> Source locations properly nested within each other

Sounds like some nice coding on a lazy Sunday - is it okay if i do it at some point? 

> ??? invariant about synthetic locations.

I can't imagine that we can do much better than `don't show these but still explore them` with the current amount of desugaring
